common/src/main/java/org/conscrypt/OkHostnameVerifier.java

```diff
@@ -0,0 +1,284 @@
+
+    /**
+     * Quick and dirty pattern to differentiate IP addresses from hostnames. This
+     * is an approximation of Android's private InetAddress#isNumeric API.
+     *
+     * <p>This matches IPv6 addresses as a hex string containing at least one
+     * colon, and possibly including dots after the first colon. It matches IPv4
+     * addresses as strings containing only decimal digits and dots. This pattern
+     * matches strings like "a:.23" and "54" that are neither IP addresses nor
+     * hostnames; they will be verified as IP addresses (which is a more strict
+     * verification).
+     */
+    private static final Pattern VERIFY_AS_IP_ADDRESS = Pattern.compile(
+            "([0-9a-fA-F]*:[0-9a-fA-F:.]*)|([\\d.]+)");
+
+    private static final int ALT_DNS_NAME = 2;
+    private static final int ALT_IPA_NAME = 7;
+
+    // BEGIN Android-changed: Add a mode which disallows top-level domain wildcards. b/144694112
+    // private OkHostnameVerifier() {
+    // }
+    private final boolean strictWildcardMode;
+
+    private OkHostnameVerifier(boolean strictWildcardMode) {
+        this.strictWildcardMode = strictWildcardMode;
+    }
+
+    public static OkHostnameVerifier strictInstance() {
+        return new OkHostnameVerifier(true);
+    }
+    // END Android-changed: Add a mode which disallows top-level domain wildcards. b/144694112
+
+    @Override
+    public boolean verify(X509Certificate[] certs, String host, SSLSession session) {
+        if (certs.length > 0) {
+            return verify(host, certs[0]);
+        } else {
+            try {
+                Certificate[] certificates = session.getPeerCertificates();
+                return verify(host, (X509Certificate) certificates[0]);
+            } catch (SSLException e) {
+                return false;
+            }
+        }
+    }
+
+    public boolean verify(String host, X509Certificate certificate) {
+        return verifyAsIpAddress(host)
+                ? verifyIpAddress(host, certificate)
+                : verifyHostName(host, certificate);
+    }
+
+    static boolean verifyAsIpAddress(String host) {
+        return VERIFY_AS_IP_ADDRESS.matcher(host).matches();
+    }
+
+    /**
+     * Returns true if {@code certificate} matches {@code ipAddress}.
+     */
+    private boolean verifyIpAddress(String ipAddress, X509Certificate certificate) {
+        List<String> altNames = getSubjectAltNames(certificate, ALT_IPA_NAME);
+        for (int i = 0, size = altNames.size(); i < size; i++) {
+            if (ipAddress.equalsIgnoreCase(altNames.get(i))) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Returns true if {@code certificate} matches {@code hostName}.
+     */
+    private boolean verifyHostName(String hostName, X509Certificate certificate) {
+        hostName = hostName.toLowerCase(Locale.US);
+        boolean hasDns = false;
```

Comment:
Let's just suppress these warnings for now and we can tidy the Android change markers later.

`@SuppressWarnings("UnusedVariable")`